### PR TITLE
Push redirect logic to the redirect page model

### PR DESCRIPTION
**Pull request by benbalter**
I did exactly what you're not supposed to do, but I did it, so I'm hoping we could still use it.

I went to tackle https://github.com/jekyll/jekyll-redirect-from/issues/125, but realized the way things were architechted, it'd be a bit tricky to do, since so much of the redirect logic was in the generator model, which was procedural and already large.

This all began when I opened Atom, and Rubocop started complaining about method length and complexity. So I started to add Rubocop, but in order to get `Generator#render` small enough, I needed to keep pushing logic into smaller methods, and it felt very un-OO to be passing the document/page around.

Then I thought, that instead, the RedirectPage should know all the things it needs to know to build itself. That's how object-oriented programming is supposed to work! So I began to pull the redirect logic out of the generator (which should concern itself with looping through things), and pushed it into the RedirectPage model.

That's all a long way to say that this embarrassingly large PR does a few things, while (theoretically) maintaining the existing functionality:

* Add Rubocop, and ensure Rubocop compliance with Jekyll coding standards/style
* Require Jekyll >= 3.0 (since we're on 3.3 now, and remove a lot of 2.x logic)
* Push the `redirect_from` and `redirect_to` YAML parsing logic into the Page and Document models directly (with `#redirect_from` and `redirect_to` helper methods that return consistent output)
* Moved the template into a standard layout and render it, rather than string interpolate it
* Sites can now override the layout by simply adding a `_layout/redirect.html` to their site
* By stubbing `Page#read_yaml`, the RedirectPage uses the standard `Page#initialize` method (less custom code)
* Use Jekyll's native `absolute_url` filter to generate canonical URLs
* Use Jekyll's native permalink/output_ext logic, rather than rolling our own for `redirect_to`'s
* Removed all the `is_*?` and `has_*?` helpers from the generator
* Tests no longer build and write the entire site between each example 

From a developer benefit, the big benefit here is adherence to Jekyll style (including method length and complexity), and a more object-oriented architecture.

From an end-user perspective, the benefit is relying more heavily on Jekyll internals, meaning theoretically, more consistent behavior, and new features going forward, in addition to custom template support.

From a going forward perspective, https://github.com/jekyll/jekyll-redirect-from/issues/125 can now be as simple as `site.pages.map { |p| p.data["redirect"] }.compact.to_json`.

I believe @parkr's :car:, but /cc @pathawks for :eyes:?

Fixes https://github.com/jekyll/jekyll-redirect-from/issues/128, https://github.com/jekyll/jekyll-redirect-from/issues/130

I think it should also fix https://github.com/jekyll/jekyll-redirect-from/issues/129 and https://github.com/jekyll/jekyll-redirect-from/issues/114.

### CODE REVIEW

**Comment by parkr** [comment]
For me, as long as all the pre-existing integration-level tests pass without alteration (i.e. it still generates all the redirect_from and redirect_to pages as it did previously), then I'm happy with whatever architecture is easiest to maintain. I can't take a close look at this now, but I'll try to get to it soon. I gladly defer to @pathawks's judgement here as well, so if he thinks it's 👍 then merge and release!

File: .travis.yml [hunk 1]
```diff
@@ -9,18 +9,11 @@ matrix:
     - # GitHub Pages
       rvm: 2.1.1
```
Comment (pathawks) [inline .travis.yml]: Note to self that this should be bumped to `2.3.3` at some point.

**Comment by benbalter** [comment]
> as long as all the pre-existing integration-level tests pass without alteration

Explicitly, there are [two integration tests](https://github.com/jekyll/jekyll-redirect-from/blob/master/spec/integrations_spec.rb), which [I expanded on](https://github.com/jekyll/jekyll-redirect-from/blob/bb-refactor/spec/integrations_spec.rb).

There are also a handful of integration-ish specs (does this file exist?) in [the redirector spec](https://github.com/jekyll/jekyll-redirect-from/blob/master/spec/jekyll_redirect_from/redirector_spec.rb#L60-L106), which *I believe* I ported over,

File: lib/jekyll-redirect-from/redirect_page.rb [hunk 2]
```diff
@@ -1,42 +1,70 @@
...
+
+    # Creates a new RedirectPage instance from the doc to the given path
+    def self.redirect_to(doc, path)
+      RedirectPage.from_paths(doc.site, doc.url, path)
+    end
+
+    # Overwrite the default read_yaml method since the file doesn't exist
+    def read_yaml(_base, _name, _opts = {})
+      self.content = ""
+      self.data ||= DEFAULT_DATA.dup
```
Comment (pathawks) [inline lib/jekyll-redirect-from/redirect_page.rb]: Why `||=`? It seems like it should either be just an assignment or a merge.

When would `self.data` already be set at this point?

File: lib/jekyll-redirect-from/redirect_page.rb [hunk 3]
```diff
@@ -1,42 +1,70 @@
 # Encoding: utf-8
-
 module JekyllRedirectFrom
+  # Specialty page which implements the redirect path logic
   class RedirectPage < Jekyll::Page
-    # Initialize a new RedirectPage.
+    # Use Jekyll's native absolute_url filter
+    include Jekyll::Filters::URLFilters
```
Comment (pathawks) [inline lib/jekyll-redirect-from/redirect_page.rb]: I _really_ wish there were a better way to do this.

Maybe the relative URL logic should be broken into a utility method?

File: lib/jekyll-redirect-from/redirect_page.rb [hunk 4]
```diff
@@ -1,42 +1,70 @@
...
+        "permalink" => from,
+        "redirect"  => {
+          "from" => from,
+          "to"   => to =~ %r!^https?://! ? to : absolute_url(to)
+        }
+      })
+    end
 
-      Jekyll::Hooks.trigger :pages, :post_init, self if JekyllRedirectFrom.jekyll_3?
+    # Redirects without a trailing slash should produce extentionless files
```
Comment (pathawks) [inline lib/jekyll-redirect-from/redirect_page.rb]: This is a regression.

Some servers will not serve extensionless files as `text/html`, which causes the browser to just download the binary file instead of loading it and following the redirect.

**Comment by benbalter** [comment]
> Let's separate the extensionless HTML files discussion into a separate PR and for now continue the previous behavior.

Can you clarify? Although I personally disagree with it, I *thought* I was implementing the behavior described in the README, no? (or is that what you're suggesting?)

> Because in other circumstances the previous version would output an absolute URL, I am willing to call this a bug in the old code and say the _new_behavior is correct.

That was my thinking, especially because there's a canonical URL field, that *must* be absolute. Also glad to expose relative and absolute URL fields, but I feel like that's overly complicated what should be a decision on our part.

File: lib/jekyll-redirect-from/redirectable.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+module JekyllRedirectFrom
+  # Module which can be mixed in to documents (and pages) to provide
+  # redirect_to and redirect_from helpers
+  module Redirectable
+    # Returns a string representing the relative path or URL
+    # to which the document should be redirected
+    def redirect_to
+      if to_liquid["redirect_to"].is_a?(Array)
+        to_liquid["redirect_to"].compact.first
```
Comment (benbalter) [inline lib/jekyll-redirect-from/redirectable.rb]: Correct. `redirect_to` *can* be an array, an we absorb the complexity of YAML and just deal with it. I'm fine with that, as many users think that array-style markup with a single entry is the proper way to do things (and I'd rather add a few method calls then break their site).

File: lib/jekyll-redirect-from/redirect.html [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
+<html lang="en-US">
+  <meta charset="utf-8">
+  <title>Redirecting...</title>
```
Comment (benbalter) [inline lib/jekyll-redirect-from/redirect.html]: Not strongly opinionated, but what's the value?

File: lib/jekyll-redirect-from/redirect_page.rb (hunk 2 above)
Comment (benbalter) [inline lib/jekyll-redirect-from/redirect_page.rb]: Ahh. You are correct.

File: lib/jekyll-redirect-from/redirect_page.rb (hunk 4 above)
Comment (benbalter) [inline lib/jekyll-redirect-from/redirect_page.rb]: I agree, but was going off of what was in the README. What's the desired behavior here?

**Comment by pathawks** [comment]
> Can you clarify? Although I personally disagree with it, I thought I was implementing the behavior described in the README, no?

It seems the behavior changed for Jekyll 3.

Current tests check to make sure that files have extensions.

### [spec_helper.rb:](https://github.com/jekyll/jekyll-redirect-from/blob/ba2efbd860495c514b561d1b955a9136a5af264d/spec/spec_helper.rb#L75-L77)

```ruby
  def forced_output_ext
    JekyllRedirectFrom.jekyll_3? ? ".html" : ""
  end
```

### [redirector_spec.rb:](https://github.com/jekyll/jekyll-redirect-from/blob/ba2efbd860495c514b561d1b955a9136a5af264d/spec/jekyll_redirect_from/redirector_spec.rb#L60-L62)
```ruby
    it "generates the refresh page for the post properly" do
      expect(dest_dir("posts/23128432159832/mary-had-a-little-lamb#{forced_output_ext}")).to exist
    end
```

---

If that is changed then :+1: from me

**Comment by pathawks** [comment]
> I thought I was implementing the behavior described in the README

If you compare [the `fixture` site](https://github.com/jekyll/jekyll-redirect-from/tree/v0.11.0/spec/fixtures) built with old code vs this code, you will see the difference in behavior.

File: lib/jekyll-redirect-from/redirect.html (hunk 6 above)
Comment (pathawks) [inline lib/jekyll-redirect-from/redirect.html]: We mean to use an ellipsis. We can use an ellipsis. We should use an ellipsis.

#93

**Comment by benbalter** [comment]
> you will see the difference in behavior.

So `/foo` should be rendered as `/foo.html` or `/foo`?

**Comment by pathawks** [comment]
According to [the fixture site](https://github.com/jekyll/jekyll-redirect-from/blob/v0.11.0/spec/fixtures/_articles/redirect-me-plz.md):

```yaml
---
redirect_from: /articles/23128432159832/mary-had-a-little-lamb
---
```

is rendered as `/articles/23128432159832/mary-had-a-little-lamb.html`

_That_ is the current behavior as of **v0.11.0**.

**Comment by benbalter** [comment]
Here are the relevant tests:

* https://github.com/jekyll/jekyll-redirect-from/pull/131/files#diff-9a838acda16e45c1f6b1abe4f2589c7bR203
* https://github.com/jekyll/jekyll-redirect-from/pull/131/files#diff-9a838acda16e45c1f6b1abe4f2589c7bR140
* https://github.com/jekyll/jekyll-redirect-from/pull/131/files#diff-72269b1dc515c5f33d598f4e84c39565R51

Just want to confirm before I implement the change, this line from the README:

> Redirects including a trailing slash will generate a corresponding subdirectory containing an index.html, while **redirects without a trailing slash will generate a corresponding filename without an extension**, and without a subdirectory.

Is not true? If we're not implementing the documented behavior, is that a bug or are the docs just out of date?

Basically, it sounds like the desired behavior is that a redirect from without an extension should be given a `.html` extension, correct?

**Comment by pathawks** [comment]
Yes, it sounds like the docs and the code are in conflict.

Relevant PR: #96

**Comment by benbalter** [comment]
@pathawks take a look at 124f85a, which should now implement that logic (always append `.html` to extensionless redirects), such that `redirect_from: /articles/23128432159832/mary-had-a-little-lamb` renders as `/articles/23128432159832/mary-had-a-little-lamb.html` as tested in https://github.com/jekyll/jekyll-redirect-from/blob/124f85ae405e3dbbedd5b62cd79f7f9b284e7c94/spec/integrations_spec.rb#L51-L56.

**Comment by benbalter** [comment]
@pathawks, thanks as always, for the critical 👀.

File: jekyll-redirect-from.gemspec [hunk 7]
```diff
@@ -18,10 +18,11 @@ Gem::Specification.new do |spec|
   spec.test_files    = spec.files.grep(%r{^(test|spec|features)/})
   spec.require_paths = ["lib"]
 
-  spec.add_runtime_dependency "jekyll", ">= 2.0"
+  spec.add_runtime_dependency "jekyll", "~> 3.0"
```
Comment (pathawks) [inline jekyll-redirect-from.gemspec]: 3.3 since we're using URL filters

**Comment by pathawks** [comment]
Thank you for all of the work here @benbalter

This will make the plugin much more robust :+1:

**Comment by benbalter** [comment]
@jekyllbot: merge +major.

**Comment by parkr** [comment]
Can we make a note in the release notes about respecting a custom `redirect.html` layout? There is no mention of it in the release notes and there are several prior releases which explicitly **removed** the layouts for this file.

Thanks!

**Comment by benbalter** [comment]
> Can we make a note in the release notes about respecting a custom redirect.html layout? There is no mention of it in the release notes

https://github.com/jekyll/jekyll-redirect-from/releases/tag/v0.12.0 had "Support for custom redirect templates" as the first major change. Updated to "Use `_layouts/redirect.html` as the redirect template if present in the site source"
